### PR TITLE
chore: configuration as code

### LINKED ISSUE #2508: Configuration as code
### Overview

We now have four build environments

* `develop`
* `release`
* `app`
* `private`

It's tedious and error prone to manually update these variables, and the only source of truth is the Fleek UI. Configuration should be managed as code.

This also gets us a step closer to managing turning on flags via PR, and including them in GitHub release notes for traceability.

### References and additional details

* Create a `base.env` file with common figuration
* Create `(local|develop|release|app|private).env` files with environment specific config that extend `base.env`
* `sample.env` should match `local.env` or similar
* Create a yarn script that can run within the fleek environment that takes a configurable argument parameter, e.g. `yarn set-env-vars release`. If an environment variables **exists** within the environment already, don't load it - env vars set via the Fleek GUI should take precedence over code, such that we can turn a flag on or off in an emergency or to time with a marketing release.
* Update the Fleek build commands for each environment to invoke the script with the respective environment prior to the build.

### Acceptance Criteria

* We have a base configuration that's common across all environments
* We can manage and version control configuration as code
* We can override configuration via the Fleek GUI
* All Fleek environments are configured correctly
* Redundant environment variables are removed from each environment

### Need By Date

_No response_

### Screenshots/Mockups

_No response_

### Estimated effort

4 hours

### PR DESCRIPTION
**Pull request by 0xdef1cafe**
## Description

**please do not merge until we go through each env with a fine tooth comb and update all hosted environments**

* proof this works https://still-sea-0314.on.fleek.co/#/demo
* the environment variables that environment is using https://still-sea-0314.on.fleek.co/env.json

this allows us to generate a `.env` file per environment by running `yarn env <dev|develop|app|private>` before the build to bootstrap the environment variables

this allows us to

* version control configuration
* establish a base environment configuration and extend base configuration per environment 
* maintain reproducible builds, as this only bootstraps environment variables
* enable operations to create PRs to set flags (to be approved by engineering) and releases to production in a self-service fashion

## Notice

- [x] Have you followed the guidelines in our [Contributing]('https://github.com/shapeshift/web/CONTRIBUTING.md) guide?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/shapeshift/web/pulls) for the same update/change?

## Pull Request Type

- [ ] :bug: Bug fix (Non-breaking Change: Fixes an issue)
- [x] :hammer_and_wrench: Chore (Non-breaking Change: Doc updates, pkg upgrades, typos, etc..)
- [ ] :nail_care: New Feature (Breaking/Non-breaking Change)

## Issue (if applicable)

closes #2508

## Risk

* bigly after we merge - we could mess up production environments by configuring them incorrectly

## Testing

**for engineering only - not ops**

* see the [temporary fleek environment here](https://app.fleek.co/#/sites/still-sea-0314/overview?accountId=73494fbc-836e-4cc8-b821-51c40f28864b) 
  * note the build command `
apt update && apt --yes install curl wget git && npm i -g n && n 16.14.2 && yarn install --frozen-lockfile && yarn env app && yarn build` 
  * **note the `yarn env app` portion**
* **note - destructive commands follow** 
  * run `yarn env dev` and see your `.env` get overwritten
  * run `yarn env app && yarn build && serve -s build -l 4000` to run a built production copy of the app

### CODE REVIEW

File: sample.env [hunk 1]
```diff
@@ -56,8 +56,6 @@ REACT_APP_GEM_ASSET_LOGO=https://gem-widgets-assets.s3-us-west-2.amazonaws.com/c
 REACT_APP_GEM_ENV=production
 REACT_APP_GEM_API_KEY=bb4164a72246dae1e03010d664d6cdae4e19b2554de02e3bf6c3cd30aa7e359e
 
-# IP/Hostname used for connecting external devices to localhost
```
Comment (0xdef1cafe) [inline sample.env:59]: should never have been committed, this was for walletconnect testing locally

File: .gitignore [hunk 2]
```diff
@@ -14,10 +14,6 @@
 # misc
 .DS_Store
 .env
-.env.local
-.env.development.local
-.env.test.local
-.env.production.local
```
Comment (0xdef1cafe) [inline .gitignore:20]: cruft

File: scripts/bootstrap-env.ts [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+// the release environment uses the app configuration
+const VALID_ENVIRONMENTS = ['local', 'develop', 'app', 'private'] as const
+type Environment = typeof VALID_ENVIRONMENTS[number]
+
+const getSerializedEnvVars = (environment: Environment) => {
+  console.log(`Using environment variables for ${environment} environment`)
+
+  const envVars = Object.assign(
+    {},
+    dotenv.parse(readFileSync('.env.base')), // always load the base config first, path is relative to root of repo
```
Comment (gomesalexandre) [inline scripts/bootstrap-env.ts:16]: `fs.promises.readFile` ?

File: scripts/bootstrap-env.ts [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+  assert(VALID_ENVIRONMENTS.includes(specifiedEnvironment), 'invalid environment')
+  return specifiedEnvironment
+}
+
+const exportDotEnvFile = (serialiazedEnvVars: string) => {
+  console.log(serialiazedEnvVars)
+  writeFileSync('.env', serialiazedEnvVars) // write out new .env
+}
+
+const main = () => exportDotEnvFile(getSerializedEnvVars(getSpecifiedEnvironment()))
```
Comment (gomesalexandre) [inline scripts/bootstrap-env.ts:38]: `lodash/flow`to make it a LTR pipe, or disable lint here and make it a multiline, closure-like expression?

**Review by gomesalexandre — COMMENTED** [review]
First pass while this is in draft, tested with `yarn env local` and `yarn env app` locally and got the right env generated - also tried to run the app as a sanity check which seemed to look good overall

File: scripts/bootstrap-env.ts (hunk 4 above)
Comment (0xdef1cafe) [reply scripts/bootstrap-env.ts]: i like the cut of your jib

**Review by gomesalexandre — APPROVED** [review]
Tested locally, seen it working in Fleek, and gave a try to `yarn && yarn env local && local-ci` which is happy

🍬

File: scripts/bootstrap-env.ts (hunk 4 above)
Comment (0xApotheosis) [reply scripts/bootstrap-env.ts]: Damn, `flow` is nice.

**Pull request merged**